packages/@aws-cdk/example-construct-library/lib/example-resource.ts

```diff
@@ -0,0 +1,501 @@
+/*
+ * We always import other construct libraries entirely with a prefix -
+ * we never import individual classes from them without a qualifier
+ * (the prefix makes it more obvious where a given dependency comes from,
+ * and prevents conflicting names causing issues).
+ * Our linter also enforces ES6-style imports -
+ * we don't use TypeScript's import a = require('a') imports.
+ */
+import * as cloudformation from '@aws-cdk/aws-cloudformation';
+import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as events from '@aws-cdk/aws-events';
+import * as iam from '@aws-cdk/aws-iam';
+import * as s3 from '@aws-cdk/aws-s3';
+import * as core from '@aws-cdk/core';
+// for files that are part of this package, we do import individual classes or functions
+import { exampleResourceArnComponents } from './example-resource-utils';
+
+/**
+ * The interface that represents the ExampleResource resource.
+ * We always use an interface, because each L2 resource type in the CDK can occur in two aspects:
+ *
+ *   1. It can be a resource that's created and managed by the CDK.
+ *     Those resources are represented by the class with the name identical to the resource -
+ *     {@link ExampleResource} in our case, which implements {@link IExampleResource}.
+ *   2. It can be a resource that exists already, and is not managed by the CDK code,
+ *     but needs to be referenced in your infrastructure definition code.
+ *     Those kinds of instances are returned from static `fromXyz(Name/Arn/Attributes)` methods -
+ *     in our case, the {@link ExampleResource.fromExampleResourceName} method.
+ *     In general, those kinds of resources do not allow any sort of mutating operations to be performed on them
+ *     (the exception is when they can be changed by creating a different resource -
+ *     IAM Roles, which you can attach multiple IAM Policies to,
+ *     are the canonical example of this sort of resource),
+ *     as they are not part of the CloudFormation stack that is created by the CDK.
+ *
+ *  So, an interface like {@link IExampleResource} represents a resource that *might* be mutable,
+ *  while the {@link ExampleResource} class represents a resource that definitely is mutable.
+ *  In almost all cases, when you're using this resource as the type of a property in a different resource definition,
+ *  you want to use {@link IExampleResource} as the type, not {@link ExampleResource}.
```

Comment:
Not clear how the first sentence connects to the second.

Maybe just simplify and say - use the interface everywhere that the resource needs to be referenced.

Reply:
Done.